### DISCUSSION TITLE
SkiaShap SKCanvasView within a ScrollView canceling Touch Event

### DISCUSSION THREAD

**Discussion by xXComanderXx** [dotnet/maui#18758] [Q&A]
We are currently integrating a SkiaShap SKCanvasView within a maui ScrollView. However, when interacting with the Skia object by touching inside it and then moving the touch outside while still holding it down, the scroll view gets triggered, causing unintended scrolling and the skia touch event is canceled. Is there a straightforward method to prevent this behavior in such cases?

**Comment by Yannikk1996** [comment]
I am encountering the same issue. Do you find a solution?

**Comment by Friedrich78** [comment]
Same issue here at Android Maui platform

**Reply by Yannikk1996** [reply]
I solved it using an derived scrollview class. Just disable scrolling when skia gets an touch event and enable scrolling again when touch is released.
Heres the code:

public partial class SkiaScrollView : ScrollView
{

    public SkiaScrollView()
	{
        Microsoft.Maui.Handlers.ScrollViewHandler.Mapper.AppendToMapping("SkiaScrollView", (handler, view) =>
        {
            if (view is SkiaScrollView)
            {
#if ANDROID
                ((SkiaScrollView)view).ScrollviewAndroid = handler.PlatformView;
                    
#endif
#if IOS
                ((SkiaScrollView)view).ScrollviewIOS = handler.PlatformView;

#endif
            }
        });
        InitializeComponent();

    }

#if IOS
        public UIKit.UIScrollView ScrollviewIOS = null;
#endif

#if ANDROID
        public Microsoft.Maui.Platform.MauiScrollView ScrollviewAndroid = null;
#endif
    public void EnableScrolling(bool Enable)
    {
#if IOS
        if (ScrollviewIOS != null)
        {
            ScrollviewIOS.ScrollEnabled = Enable;
        }
#endif

#if ANDROID
if(ScrollviewAndroid != null)
{ScrollviewAndroid.RequestDisallowInterceptTouchEvent(true); 
}
#endif
    }

}

**Comment by Friedrich78** [comment]
Hi Yannikk1996
I implemented it like suggested from you (a bit different since I have a csproj project per target) for Android only. Unfortunately it is not working even If I set RequestDisallowInterceptTouchEvent(true) right at the beginning.

My code in the Android project (UnifiedHmiScrollView class is defined in a class library and derived from ScrollView):

                ScrollViewHandler.Mapper.AppendToMapping(nameof(UnifiedHmiScrollView.IsScrollingDisabled), (handler, view) =>
                {
                    var platformView = handler.PlatformView;
                    var virtualView = handler.VirtualView as UnifiedHmiScrollView;

                    if (virtualView != null)
                    {
                        platformView.RequestDisallowInterceptTouchEvent(virtualView.IsScrollingDisabled);
                    }
                });

Interestingly the Touchdown and some Move events arrive in the SKCanvasView but suddenly after maybe 1 second a Cancelled arrives. If I remove the ScrollView from the parent hierachy then everything works fine.

Note my control hierarchy looks like this

 -> ScrollView
----- -> ContentView
----------- -> ContentView
------------------ -> ContentView
------------------------- -> SKCanvasView

Do you have any ideas?

**Comment by imuller** [comment]
Simplified the code of @Yannikk1996 a little bit:

```
public class SkiaScrollView : ScrollView
{
#if IOS
    private UIKit.UIScrollView _scrollViewIOS;
#elif ANDROID
    private Microsoft.Maui.Platform.MauiScrollView _scrollViewAndroid;
#endif
    
    public SkiaScrollView()
    {
        Microsoft.Maui.Handlers.ScrollViewHandler.Mapper.AppendToMapping("SkiaScrollView", (handler, view) =>
        {
            if (view is not SkiaScrollView)
                return;
#if ANDROID
            _scrollViewAndroid = handler.PlatformView;
#elif IOS
            _scrollViewIOS = handler.PlatformView;
#endif
        });
    }
    
    /// <summary>
    /// Enable or disable scrolling of the scrollview
    /// </summary>
    /// <param name="enable"></param>
    public void EnableScrolling(bool enable)
    {
#if IOS
        if (_scrollViewIOS != null)
            _scrollViewIOS.ScrollEnabled = enable;
#elif ANDROID
        _scrollViewAndroid?.RequestDisallowInterceptTouchEvent(!enable);
#endif
    }
}
```

**Comment by Friedrich78** [comment]
I found out the EnableScrolling method works only for ScrollView.Orientation="Vertical" for Android. We need ScrollView.Orientation="Both" and then we are facing the issues below. 

Example for xaml to reproduce:

```
<ContentPage xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
             xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
             xmlns:skScrollView="clr-namespace:SKScrollView1"
             x:Class="SKScrollView1.MainPage">
    <skScrollView:SkiaScrollView x:Name="ScrollView" Orientation="Both">
        <AbsoluteLayout>
            <skScrollView:MyRectangle AbsoluteLayout.LayoutBounds="100, 100, 100, 100"/>

            <skScrollView:MyRectangle AbsoluteLayout.LayoutBounds="300, 100, 100, 100"/>

            <skScrollView:MyRectangle AbsoluteLayout.LayoutBounds="350, 600, 100, 100"/>

            <skScrollView:MyRectangle AbsoluteLayout.LayoutBounds="300, 800, 100, 100"/>
        </AbsoluteLayout>
    </skScrollView:SkiaScrollView>
</ContentPage>

```

Code for MyRectangle:
 ```
public class MyRectangle : SKCanvasView
 {
     public MyRectangle()
     {
         this.PaintSurface += MyRectangle_PaintSurface;
         this.EnableTouchEvents = true;
     }

     protected override void OnTouch(SKTouchEventArgs e)
     {
         if (e.ActionType == SKTouchAction.Pressed)
         {
             EnableScrolling(false);
         }
         else if (e.ActionType == SKTouchAction.Released)
         {
             EnableScrolling(true);
         }

         e.Handled = true;

         base.OnTouch(e);

         Debug.WriteLine("MyRectangle.OnTouch: " + e.ActionType);
     }

     private void MyRectangle_PaintSurface(object? sender, SkiaSharp.Views.Maui.SKPaintSurfaceEventArgs e)
     {
         SKSurface surface = e.Surface;
         surface.Canvas.Clear();

         var rect = new SKRect(0, 0, CanvasSize.Width, CanvasSize.Height);
         var paint = new SKPaint()
         {
             Style = SKPaintStyle.Fill,
             Color = SKColors.Aqua,
             StrokeWidth = 0,
             StrokeCap = SKStrokeCap.Square,
         };

         surface.Canvas.DrawRect(rect, paint);
     }

     private void EnableScrolling(bool enable)
     {
         var parent = (SkiaScrollView)this.Parent.Parent;
         parent.EnableScrolling(enable);
     }
 }
```
